File: frontend/src/services/supabaseClient.js
import { createClient } from '@supabase/supabase-js';
import { getPatientSessionId } from '../utils/session';

// Direct Supabase Credentials
const DEFAULT_SUPABASE_URL = "https://swxwtlqvpmbrzwjbvmva.supabase.co";
const DEFAULT_SUPABASE_KEY = "sb_publishable_LkPBGMkAOATPB2qUxBz0cA_2egzKInV";

const SUPABASE_URL = import.meta.env.VITE_SUPABASE_URL || DEFAULT_SUPABASE_URL;
const SUPABASE_KEY = import.meta.env.VITE_SUPABASE_KEY || DEFAULT_SUPABASE_KEY;

let supabase = null;

if (SUPABASE_URL && SUPABASE_KEY) {
  try {
    supabase = createClient(SUPABASE_URL, SUPABASE_KEY);
    console.log('[Supabase Client] Initialized with URL:', SUPABASE_URL);
  } catch (e) {
    console.warn("Supabase init error:", e);
  }
}

export async function savePredictionToSupabase(labData, prediction) {
  if (!supabase) {
    console.warn('[Supabase] Client not initialized!');
    return null;
  }

  const sessionId = getPatientSessionId();
  const reportId = `CDSS_Report_${Date.now()}`;

  const row = {
    id: reportId,
    patient_session_id: sessionId,
    timestamp: new Date().toISOString().replace('T', ' ').substring(0, 19),
    patient_age: Number(labData.age) || 45,
    patient_gender: labData.gender || 'Female',
    hba1c: Number(labData.hba1c) || 5.8,
    fasting_glucose: Number(labData.fasting_glucose) || 105.0,
    random_glucose: Number(labData.random_glucose) || 140.0,
    predicted_class: prediction.predicted_class,
    confidence_score: prediction.confidence_score,
    severity_index: prediction.severity_index,
    storage_provider: 'Supabase Cloud Realtime'
  };

  try {
    const { data, error } = await supabase
      .from('patient_reports')
      .insert([row]);

    if (error) {
      console.error('[Supabase Insert Error]:', error);
      return null;
    }

    console.log('[Supabase] Successfully saved record to patient_reports table:', row);
    return row;
  } catch (err) {
    console.error('[Supabase Exception]:', err);
    return null;
  }
}

export async function fetchSessionReportsFromSupabase(sessionId = getPatientSessionId()) {
  if (!supabase) return [];

  try {
    // Filter strictly by patient_session_id to prevent cross-patient report viewing
    const { data, error } = await supabase
      .from('patient_reports')
      .select('*')
      .eq('patient_session_id', sessionId)
      .order('created_at', { ascending: false });

    if (error) {
      console.error('[Supabase Fetch Error]:', error);
      return [];
    }

    return data || [];
  } catch (err) {
    console.error('[Supabase Fetch Exception]:', err);
    return [];
  }
}

File: src/supabase_client.py
import os
import json
import time
from typing import List, Dict, Any

DEFAULT_SUPABASE_URL = "https://swxwtlqvpmbrzwjbvmva.supabase.co"
DEFAULT_SUPABASE_KEY = "sb_publishable_LkPBGMkAOATPB2qUxBz0cA_2egzKInV"

SUPABASE_URL = os.getenv("SUPABASE_URL", DEFAULT_SUPABASE_URL)
SUPABASE_KEY = os.getenv("SUPABASE_KEY", os.getenv("SUPABASE_SECRET_KEY", os.getenv("VITE_SUPABASE_KEY", DEFAULT_SUPABASE_KEY)))

def is_supabase_configured() -> bool:
    return bool(SUPABASE_URL and SUPABASE_KEY)

def safe_int(val, default=45) -> int:
    try:
        if val is None or val == "":
            return default
        return int(val)
    except (ValueError, TypeError):
        return default

def safe_float(val, default=5.8) -> float:
    try:
        if val is None or val == "":
            return default
        return float(val)
    except (ValueError, TypeError):
        return default

def save_report_to_supabase(
    patient_data: Dict[str, Any],
    prediction: Dict[str, Any],
    generated_pdf_bytes: bytes = None,
    uploaded_pdf_bytes: bytes = None,
    patient_session_id: str = "session_default"
) -> Dict[str, Any]:
    """
    Saves report metadata, raw uploaded PDF, and generated clinical PDF to Supabase Database & Storage.
    Safely parses numeric inputs to prevent type errors.
    """
    report_id = f"CDSS_Report_{int(time.time())}"
    
    # Core row matching user's Supabase table schema with safe parsing & patient_session_id isolation
    db_row = {
        "id": report_id,
        "patient_session_id": patient_session_id,
        "timestamp": time.strftime("%Y-%m-%d %H:%M:%S"),
        "patient_age": safe_int(patient_data.get("age"), 45),
        "patient_gender": str(patient_data.get("gender") or "Female"),
        "hba1c": safe_float(patient_data.get("hba1c"), 5.8),
        "fasting_glucose": safe_float(patient_data.get("fasting_glucose"), 105.0),
        "predicted_class": str(prediction.get("predicted_class") or "Healthy")
    }

    if is_supabase_configured():
        try:
            from supabase import create_client
            supabase = create_client(SUPABASE_URL, SUPABASE_KEY)

            # Insert metadata into 'patient_reports' table
            res = supabase.table("patient_reports").insert(db_row).execute()
            print(f"[Supabase SUCCESS] Inserted row into patient_reports table: {db_row}")

        except Exception as e:
            print(f"[Supabase Exception]: {e}")

    return db_row

def fetch_reports_from_supabase(patient_session_id: str = None) -> List[Dict[str, Any]]:
    """Fetches patient reports from Supabase Table Editor safely with patient_session_id isolation."""
    if is_supabase_configured():
        try:
            from supabase import create_client
            supabase = create_client(SUPABASE_URL, SUPABASE_KEY)
            query = supabase.table("patient_reports").select("*")
            if patient_session_id and patient_session_id != "ALL":
                query = query.eq("patient_session_id", patient_session_id)
            response = query.order("created_at", desc=True).execute()
            if response.data:
                return response.data
        except Exception as e:
            print(f"[Supabase Query Exception]: {e}")
            
    return []
